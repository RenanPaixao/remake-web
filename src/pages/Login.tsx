import { Button, Center, FormControl, FormErrorMessage, FormLabel, Input, Text } from '@chakra-ui/react'
import { Field, Form, Formik, FormikTouched, FormikErrors } from 'formik'
import * as Yup from 'yup'
import React from 'react'
import { supabase } from '../utils/supabase.ts'
import { Link, useNavigate } from 'react-router-dom'
import { useToast } from '@chakra-ui/react'

interface FormValues {
  email: string
  password: string
}

const validationSchema = Yup.object({
  email: Yup.string().email('Invalid email address').required('Required'),
  password: Yup.string().required('Required').min(8, 'Password must be at least 8 characters')
})

export const Login = (): React.JSX.Element => {
  const toast = useToast()
  const navigate = useNavigate()
  /**
   * Submit the login form.
   * @param values
   */
  async function submitForm(values: FormValues): Promise<void> {
    const { error } = await supabase.auth.signInWithPassword({
      email: values.email,
      password: values.password
    })

    if (error) {
      toast({
        title: 'Error',
        description: error.message,
        status: 'error',
        isClosable: true
      })
      console.error(error)
      return
    }

    toast({
      title: 'Success',
      description: 'You have successfully logged in.',
      status: 'success',
      isClosable: true
    })

    navigate('/')
  }

  /**
   * Check if a field has an error.
   *
   * @param name - The name of the field.
   * @param touched - The object with all touched fields.
   * @param errors - The object with all errors.
   */
  function hasError(name: keyof FormValues, touched: FormikTouched<FormValues>, errors: FormikErrors<FormValues>): boolean {
    return !!touched[name] && !!errors[name]
  }

  return <Center height='100vh' flexDirection='column'>
    <Formik initialValues={{
      email: '',
      password: ''
    }}
    onSubmit={submitForm}
    validationSchema={validationSchema}
    >
      {({ isSubmitting, errors, touched }) => (
        <Form>
          <Field name="email">
            {({ field }: any) => (
              <FormControl isInvalid={hasError('email', touched, errors)} isRequired>
                <FormLabel htmlFor="email">Email</FormLabel>
                <Input {...field} id="email" placeholder="Email" />
                <FormErrorMessage>{errors.email}</FormErrorMessage>
              </FormControl>
            )}
          </Field>

          <Field name="password">
            {({ field }: any) => (
              <FormControl mt={4} isInvalid={hasError('password', touched, errors)} isRequired>
                <FormLabel htmlFor="password">Password</FormLabel>
                <Input type='password' {...field} id="password" placeholder="Password" />
                <FormErrorMessage>{errors.password}</FormErrorMessage>
              </FormControl>
            )}
          </Field >

          <Button
            colorScheme='blue'
            color="white"
            mt={4}
            isLoading={isSubmitting}
            isDisabled={isSubmitting || Object.keys(errors).length > 0 }
            type="submit"
          >
            Login
          </Button>
        </Form>
      )}
    </Formik>
    <Text mt='2rem' fontSize='0.9rem'>Don't have an account?
      <Link to='/sign-up'>
        <strong> Register here!</strong>
      </Link>
    </Text>
    <Text fontSize='0.9rem'>
      <Link to='/forgot-password'>
        <strong>Forgot password?</strong>
      </Link>
    </Text>
  </Center>
}
