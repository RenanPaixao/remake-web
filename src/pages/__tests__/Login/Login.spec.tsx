import { Login } from '../../Login.tsx'
import { fireEvent } from '@testing-library/dom'
import { act } from '@testing-library/react'
import { expect } from 'vitest'
import { customRender, testingHistory } from '../../../../tests/test-utils.tsx'

type Navigator = {
  geolocation: {
    getCurrentPosition: (success: () => void) => void
  }
}

(navigator as Navigator).geolocation  = {
  getCurrentPosition: vi.fn().mockImplementation(success =>
    Promise.resolve(
      success({
        coords: {
          latitude: 0,
          longitude: 0
        }
      })
    )
  )
}

const mockPushHistory = vi.fn()
testingHistory.push = mockPushHistory

describe('Login', () => {
  beforeEach(() => {
    vi.mock('react-router-dom', async importOriginal => {
      const actual: object = await importOriginal()
      return {
        ...actual,
        useNavigate: () => vi.fn()
      }
    })
  })

  it('should render', async () => {
    const { container } = customRender(<Login/>)

    expect(container).toMatchSnapshot()
  })

  it('should redirect correctly to sign-up and forgot-password', async () => {
    const { getByRole } = customRender(<Login/>)

    const signUpLink = getByRole('link', { name: /register here!/i })
    const forgotPasswordLink = getByRole('link', { name: /forgot your password?/i })

    expect(signUpLink).toHaveAttribute('href', '/sign-up')
    expect(forgotPasswordLink).toHaveAttribute('href', '/forgot-password')

    act(()=> {
      fireEvent.click(signUpLink)
      fireEvent.click(forgotPasswordLink)
    })

    expect(mockPushHistory).toHaveBeenCalledTimes(2)
    expect(mockPushHistory.mock.calls[0]).toEqual(expect.arrayContaining([{
      pathname: '/sign-up',
      search: '',
      hash: ''
    }]))

    expect(mockPushHistory.mock.calls[1]).toEqual(expect.arrayContaining([
      {
        pathname: '/forgot-password',
        search: '',
        hash: ''
      }
    ]))
  })

  it('should assert that location is being saved on sessionStorage', async () => {
    customRender(<Login/>)

    expect(sessionStorage.getItem('location')).toEqual(JSON.stringify({ latitude:0,longitude:0 }))
  })
})
