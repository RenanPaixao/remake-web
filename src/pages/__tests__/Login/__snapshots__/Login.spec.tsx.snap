// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > should render 1`] = `
<div>
  <div
    class="css-b93e9j"
  >
    <img
      alt="Remake logo"
      class="chakra-image css-ibmrth"
      src="/src/assets/logo.webp"
    />
    <form
      action="#"
    >
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="email"
          id="field-:r0:-label"
        >
          Email
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-1tfjd1n"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-required="true"
          class="chakra-input css-1cjy4zv"
          id="email"
          name="email"
          placeholder="Email"
          required=""
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1mn9ptq"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="password"
          id="field-:r1:-label"
        >
          Password
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-1tfjd1n"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-required="true"
          class="chakra-input css-1cjy4zv"
          id="password"
          name="password"
          placeholder="Password"
          required=""
          type="password"
          value=""
        />
      </div>
      <button
        class="chakra-button css-13nworv"
        role="button"
        type="submit"
      >
        Login
      </button>
    </form>
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-fckmjt"
      >
        Don't have an account?
        <a
          href="/sign-up"
          role="link"
        >
          <strong>
             
            Register here!
          </strong>
        </a>
      </p>
      <p
        class="chakra-text css-1uno72p"
      >
        <a
          href="/forgot-password"
          role="link"
        >
          <strong>
            Forgot your password?
          </strong>
        </a>
      </p>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
